### PR TITLE
update castle sdk remove deprecated fields

### PR DESCRIPTION
**Pull request by pansbro12**
Updating the request sent to castle to include the created_at timestamp, context [here](https://utilitywarehouse.slack.com/archives/C04UATB9796/p1722430942888279), and fix general api drift.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -118,7 +125,13 @@ func (c *Castle) sendCall(ctx context.Context, r *castleAPIRequest, url string)
 	}()
 
 	b := new(bytes.Buffer)
-	err = json.NewEncoder(b).Encode(r)
+
+	switch request := r.(type) {
+	case *castleRiskAPIRequest:
+		err = json.NewEncoder(b).Encode(request)
+	case *castleFilterAPIRequest:
+		err = json.NewEncoder(b).Encode(request)
```
Comment (matiit) [inline client.go:133]:
```suggestion
	case *castleRiskAPIRequest:
	case *castleFilterAPIRequest:
		err = json.NewEncoder(b).Encode(request)
```

Comment (matiit) [reply client.go:133]: Can do just that?

Comment (denisnikulin) [reply client.go:133]: or commas https://go.dev/wiki/Switch#multiple-cases

**Review by Jarota — COMMENTED** [review]
Now that Risk and Filter aren't as similar as they were before, I wonder if it makes sense to split the code into a `risk.go` file and a `filter.go` file, where the relevant types are defined next to the corresponding API methods?

File: client.go [hunk 2]
```diff
@@ -92,19 +98,20 @@ func (c *Castle) Risk(ctx context.Context, req *Request) (RecommendedAction, err
...
-	r := &castleAPIRequest{
+	r := &castleRiskAPIRequest{
 		Type:         req.Event.EventType,
 		Name:         req.Event.Name,
 		Status:       req.Event.EventStatus,
 		RequestToken: req.Context.RequestToken,
 		User:         req.User,
 		Context:      req.Context,
 		Properties:   req.Properties,
+		CreatedAt:    time.Now(),
```
[the remark this answers is not in the document]
Comment (pansbro12) [reply client.go:109]: on yeah your right, I had kind of forgotten where this code is invoked

File: client.go (hunk 1 above)
Comment (pansbro12) [reply client.go:133]: have chosen commas for now, had a look at generics but I honestly haven't quite grasped them yet so I may have missed something.

Comment (Jarota) [reply client.go:133]: https://github.com/utilitywarehouse/castle-go/pull/27/files

^ for illustration purposes

File: client.go [hunk 3]
```diff
@@ -118,7 +124,11 @@ func (c *Castle) sendCall(ctx context.Context, r *castleAPIRequest, url string)
 	}()
 
 	b := new(bytes.Buffer)
-	err = json.NewEncoder(b).Encode(r)
+
+	switch request := r.(type) {
+	case *castleFilterAPIRequest, *castleRiskAPIRequest:
```
Comment (Jarota) [inline client.go:129]: Can add a default case that errors to be safe

File: client.go [hunk 4]
```diff
@@ -92,19 +97,20 @@ func (c *Castle) Risk(ctx context.Context, req *Request) (RecommendedAction, err
...
-	r := &castleAPIRequest{
+	r := &castleRiskAPIRequest{
 		Type:         req.Event.EventType,
 		Name:         req.Event.Name,
 		Status:       req.Event.EventStatus,
 		RequestToken: req.Context.RequestToken,
 		User:         req.User,
 		Context:      req.Context,
 		Properties:   req.Properties,
+		CreatedAt:    req.CreatedAt,
```
Comment (Jarota) [inline client.go:108]: Possibly worth defaulting to `time.Now()` if the created at isn't set in the request, that will preserve existing behaviour and avoid a scenario where we update to latest `castle-go` without updating the Risk calls to set created at?

File: client.go [hunk 5]
```diff
@@ -75,6 +76,10 @@ func (c *Castle) Filter(ctx context.Context, req *Request) (RecommendedAction, e
 		Email:    req.User.Email,
 		Username: req.User.Name,
 	}
+	createdAt := req.CreatedAt
+	if req.CreatedAt.IsZero() {
```
Comment (Jarota) [inline client.go:80]:
```suggestion
	createdAt := req.CreatedAt
	if createdAt.IsZero() {
```

File: client.go [hunk 6]
```diff
@@ -92,19 +102,24 @@ func (c *Castle) Risk(ctx context.Context, req *Request) (RecommendedAction, err
 	if req.Context == nil {
 		return RecommendedActionNone, errors.New("request.Context cannot be nil")
 	}
-	r := &castleAPIRequest{
+	createdAt := req.CreatedAt
+	if req.CreatedAt.IsZero() {
```
Comment (Jarota) [inline client.go:106]:
```suggestion
	createdAt := req.CreatedAt
	if createdAt.IsZero() {
```

File: client.go [hunk 7]
```diff
@@ -71,14 +72,23 @@ func (c *Castle) Filter(ctx context.Context, req *Request) (RecommendedAction, e
 	if req.Context == nil {
 		return RecommendedActionNone, errors.New("request.Context cannot be nil")
 	}
-	r := &castleAPIRequest{
+	params := Params{
+		Email:    req.User.Email,
+		Username: req.User.Name,
```
Comment (Jarota) [inline client.go:77]: Is the name necessarily the username? I wonder if we're reusing the `name` field in our generic request type in a way that wasn't intended by the underlying API, or are they actually the same values?

Comment (pansbro12) [reply client.go:77]: No so i actually was just thinking about this now, I remember us using either the account number or the email as username in a service somewhere but i can't think were.

Comment (pansbro12) [reply client.go]: Not really sure whether to include this now, looking at the code from the castle-processor the username is seems to be either the email or account number depending on what the user used to sign in. The username value is contained in the request properties so I've just set that as the username, might be that the username and the email are duplicates in some cases but I don't think that will be an issue.

**Review by Jarota — APPROVED** [review]

**Pull request merged**